### PR TITLE
Improve Autoconf.

### PR DESCRIPTION
**Pull request by apteryks**
This modernizes the Autoconf configure script a bit.

### CODE REVIEW

File: configure.in [hunk 1]
```diff
@@ -1,6 +1,6 @@
 dnl Process this file with autoconf to produce a configure script.
 
-AC_PREREQ(2.58)
```
Comment (imciner2) [inline configure.in:3]: Why this large a version jump? Autoconf 2.71 was only released 2 years ago and isn't in linux distros like Centos Stream 9 or Debian 12. wxWidgets tends to be conservative in its support for tools (hence also only supporting up to C++11, and that was only increased very recently).

Comment (Apteryks) [reply configure.in:3]: I thought 2 years was plenty of time :-).  Debian 11 is now oldstable, replaced by Debian 12 Bookworm, which has 2.71.  Also, even if Autoconf 2.71 is not available on some old platform, the configure script should work the same (and it's already distributed in the source archive of wxWidgets), so unless someone wants to modify the wxWidgets Autotools-based build system on oldstable or Centos 9, that shouldn't be an issue.  What do you think?

**Comment by oneeyeman1** [comment]
@Apteryks ,

> I thought 2 years was plenty of time :-).

Not even close. ;-)
Try something like 15-20.

Windows XP support was dropped only last year, if not this one...
And so was GTK1 and Motiff.
OSX 10.6 was dropped I think 3 years ago.

So unless Vadin say this change is good - you will be asked to revert that.

Thank you.

**Comment by vadz** [comment]
Thanks for the update, I agree that we should modernize our configure script and stop using deprecated macros, but I am also not sure it's worth bumping up the requirement to be so high as we don't seem to be using anything specific to 2.71, do we? We can increase the version requirement if really needed, but I prefer not to do it without any good reason, it's not like accepting old versions does any harm.

I'd also welcome actually removing obsolete tests rather than updating them to use newer macros. E.g. one of the modifications updates the macro preceded by a comment about egcs (!) 1.1.1 (!!), i.e. a compiler obsolete since more than 25 years. It would be really better to simplify or remove this test entirely instead of putting lipstick on it...

**Comment by Apteryks** [comment]
Hello,

I can look into downgrading the version of Autoconf used, although I suppose its autoupdate doesn't know about all the tricks it currently knows and autoupdate would painfully (because of its limitations and the tricky manual fixups involved) need to be redone at some later time.  I would have preferred to bring it already to its best/most current form.

About the egcs test, I'll remove it in a subsequent commit.  Thanks for the feedback so far!  Any clue as to what failed in the appveyor CI?  It seems I cannot see the report.

**Comment by Apteryks** [comment]
Hi! The `AC_AC_PROG_CC` was a typo of mine.  It should have been `AC_PROG_CC`.  I've also gotten rid of the whitespace.  Thankfully, re-running autoupdate again doesn't reintroduce these spurious changes.

I've taken some steps to get rid of crufty M4 modules and avoiding bundling so much in Git in the last commits.

File: build/aclocal/ax_cxx_compile_stdcxx.m4 [hunk 2]
```diff
@@ -36,14 +36,14 @@
 #   Copyright (c) 2016, 2018 Krzesimir Nowak <qdlacz@gmail.com>
 #   Copyright (c) 2019 Enji Cooper <yaneurabeya@gmail.com>
 #   Copyright (c) 2020 Jason Merrill <jason@redhat.com>
-#   Copyright (c) 2021 Jörn Heusipp <osmanx@problemloesungsmaschine.de>
+#   Copyright (c) 2021 JÃ¶rn Heusipp <osmanx@problemloesungsmaschine.de>
```
Comment (vadz) [inline build/aclocal/ax_cxx_compile_stdcxx.m4:39]: This looks like an encoding problem.

File: configure.in [hunk 3]
```diff
@@ -1,6 +1,5 @@
 dnl Process this file with autoconf to produce a configure script.
-
-AC_PREREQ(2.58)
+AC_PREREQ([2.71])
```
Comment (vadz) [inline configure.in:2]: As mentioned, I think we can use 2.69 here without any other changes to this file. Could you please check that it indeed works and make this change? 2.69 is the version in Debian 11 (bullseye) and I don't see any reason to drop support for it if we don't use anything that it doesn't support.

File: configure.in [hunk 4]
```diff
@@ -3861,11 +3811,8 @@ dnl MinGW has a vswprintf with a different prototype, and
 dnl a _vsnwprintf with the correct prototype, but AC_CHECK_FUNCS
 dnl finds it even if it is not declared in some versions...
 AC_MSG_CHECKING([for _vsnwprintf])
-AC_TRY_COMPILE([#include <wchar.h>],
-               [&_vsnwprintf;],
-               [AC_MSG_RESULT([yes])
-                AC_DEFINE(HAVE__VSNWPRINTF)],
-               [AC_MSG_RESULT([no])]);
+AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[#include <wchar.h>]], [[&_vsnwprintf;]])],[AC_MSG_RESULT([yes])
```
Comment (vadz) [inline configure.in:3814]: I think the old indentation/line wrapping was more clear.

File: configure.in [hunk 5]
```diff
@@ -4333,7 +4252,7 @@ if test "$TOOLKIT" != "MSW"; then
 
         case "${host}" in
             *-*-qnx*)
-                dnl -lpthread works, i.e. AC_TRY_LINK() would pass, but results
+                dnl -lpthread works, i.e. AC_LINK_IFELSE([AC_LANG_PROGRAM([[]], [[]])],[],[]) would pass, but results
```
Comment (vadz) [inline configure.in:4255]: This autoupdated comment is still unreadable.

**Review by vadz — CHANGES_REQUESTED** [review]
Could you please explain _why_ do we need [Stop versioning M4 macros provided by Autoconf](https://github.com/wxWidgets/wxWidgets/pull/23640/commits/9ddafe360f393b9fe47e1120f4c1537fe5861a32) commit? This results in a lot of changes, which it's going to take time to review and I just don't understand what do we gain from doing it. In fact, I strongly suspect we lose something here as removing `bakefile*.m4` files will definitely make life more difficult for people who want to regenerate configure.

Concerning the result of the changes, they look mostly fine to me (except for some comments/wrapping/whitespace changes that still remain), but we definitely need to test them as they're extensive enough that it's hard to review all of them manually and most CI builds didn't run because this branch isn't mergeable -- could you please rebase it on the current master and force-push? Ideal would be to do it with the current version, and then make any more changes on the rebased branch, as this will allow to see diffs compared to this version.

TIA!

File: configure.in (hunk 3 above)
Comment (Apteryks) [reply configure.in:2]: I'll put this idea to the test, but I suspect some things *did* change, such as the AC_PROC_CC_99 macro having been made obsolete by AC_PROG_CC since Autoconf 2.70.

**Comment by Apteryks** [comment]
> _why_ do we need [Stop versioning M4 macros provided by Autoconf](https://github.com/wxWidgets/wxWidgets/pull/23640/commits/9ddafe360f393b9fe47e1120f4c1537fe5861a32) commit

It's mostly to ease maintenance; the packages such as Bakefile, Autoconf and pkg-config provide their own M4 macros; we should use these as they are sometimes updated, e.g. to avoid Autotools deprecation warnings or bug fixes.  I've tried explaining the rationale in the commit message, see: https://github.com/wxWidgets/wxWidgets/pull/23640/commits/56126a83ff8d77aa54055cc5227882c29fb28be3.

File: configure.in (hunk 3 above)
Comment (Apteryks) [reply configure.in]: There was no problem, done!

**Comment by vadz** [comment]
> > _why_ do we need [Stop versioning M4 macros provided by Autoconf](https://github.com/wxWidgets/wxWidgets/pull/23640/commits/9ddafe360f393b9fe47e1120f4c1537fe5861a32) commit
> 
> It's mostly to ease maintenance; the packages such as Bakefile, Autoconf and pkg-config provide their own M4 macros; we should use these as they are sometimes updated, e.g. to avoid Autotools deprecation warnings or for bug fixes. I've tried explaining the rationale in the commit message, see: [56126a8](https://github.com/wxWidgets/wxWidgets/commit/56126a83ff8d77aa54055cc5227882c29fb28be3).

I think you must have meant eab3075f01c75f979c18927bbc7d25f82fbd4158 but, if so, I remain rather unconvinced by this argument. Having these files has never been a burden, while not having them will almost certainly be a problem in practice, especially for bakefile which is not easily installable anywhere by now.

Sorry but I don't see any point to complicate our lives by removing these files and I really don't want to merge this commit.

**Comment by Apteryks** [comment]
> > > _why_ do we need [Stop versioning M4 macros provided by Autoconf](https://github.com/wxWidgets/wxWidgets/pull/23640/commits/9ddafe360f393b9fe47e1120f4c1537fe5861a32) commit
> > 
> > 
> > It's mostly to ease maintenance; the packages such as Bakefile, Autoconf and pkg-config provide their own M4 macros; we should use these as they are sometimes updated, e.g. to avoid Autotools deprecation warnings or for bug fixes. I've tried explaining the rationale in the commit message, see: [56126a8](https://github.com/wxWidgets/wxWidgets/commit/56126a83ff8d77aa54055cc5227882c29fb28be3).
> 
> I think you must have meant [eab3075](https://github.com/wxWidgets/wxWidgets/commit/eab3075f01c75f979c18927bbc7d25f82fbd4158) but, if so, I remain rather unconvinced by this argument. Having these files has never been a burden, while not having them will almost certainly be a problem in practice, especially for bakefile which is not easily installable anywhere by now.
> 
> Sorry but I don't see any point to complicate our lives by removing these files and I really don't want to merge this commit.

That's fair.  If it's useful to some to have these files checked in then it's a good enough reason to keep them, and I don't mind dropping this commit.  Regarding the availability of Bakefile, that's indeed problematic; for what it's worth, it's now possible to have it installed via Guix by adding the `Guix-Past` channel: https://gitlab.inria.fr/guix-hpc/guix-past/-/commit/04364fb15ab8c751006e1a05f3b8219d5a0e6410.

I've been working on these changes from a `guix shell -D wxwidgets autoconf@2.69 automake bakefile libtool catch2@2 sdl` environment.

**Comment by Apteryks** [comment]
Latest push with controversial commit dropped (stop versioning package-provided M4 macros).

**Comment by Apteryks** [comment]
> I'm sorry, I don't understand: why are some M4 files still deleted in the latest version? What's different about them?

The three M4 files deleted in the "build: Delete unused M4 macros." is because they provide macros that are not used (per git grep'ing the source), unless I've missed something.  It's dead code.

**Comment by vadz** [comment]
Sorry but you really need to look at `getservbyname`-related build failures before doing anything else with this or the related PRs, as long as Mac build is broken this can't be merged anyhow.

**Comment by Apteryks** [comment]
> Sorry but you really need to look at `getservbyname`-related build failures before doing anything else with this or the related PRs, as long as Mac build is broken this can't be merged anyhow.

That should now be fixed. Thanks for pointing what the issue was caused by. The last remaining check failing is a false positive (it complains there's a tab in a M4 file -- it's normal as it's a macro expanding to Make).

File: .gitignore [hunk 6]
```diff
@@ -11,10 +11,20 @@
...
+*.o
+*.so
+*.so.*
+.deps
+aclocal.m4
+bk-deps
+lib/wx/
+utils/ifacecheck/rungccxml.sh
+utils/wxrc/wxrc
+version-script
```
Comment (vadz) [inline .gitignore:25]: Sorry, no, we don't want to have this here. Please run the builds in a different directory.

File: build/aclocal/ax_gxx_version.m4
Comment (vadz) [inline build/aclocal/ax_gxx_version.m4:1]: Just for the record, this had been used until 4e802e6fc7 (Remove wxUSE_COMPILER_TLS and always use C++11 thread_local, 2022-10-26) but is indeed not needed any more.

File: configure.in [hunk 7]
```diff
@@ -1345,10 +1346,8 @@ version or omitting it entirely."
             #elif !defined(__IPHONE_OS_VERSION_MIN_REQUIRED)
             #error unrecognized platform
             #endif
-        ],
-        [AC_MSG_RESULT([yes])],
-        [AC_MSG_FAILURE([$error_message])]
-    )
+        ]])],[AC_MSG_RESULT([yes])],[AC_MSG_FAILURE([$error_message])
```
Comment (vadz) [inline configure.in:1349]: I'd really appreciate if you could please restore the previous indentation which is much more readable, as otherwise I'd have to do it myself.

(this applies here and in several more places below)

**Review by vadz — COMMENTED** [review]
Thanks for fixing the CI (and yes, indeed, the remaining failure is spurious, I'll exclude this `.mk` file from the check in master), but there are still several things I don't understand here.

1. Why does `aclocal.m4` need to be removed? As long as we store all the generated files in the VCS (and, for better or worse, we do it), it should stay IMO.
2. If we remove the "autohacks" stuff, wouldn't it result in configure generated under different systems be different again? Perhaps we don't care about it any longer because I personally just use `build/tools/autoconf/Dockerfile` for regenerating this anyhow, but this is not at all clear to me.

More generally, I'd really like to understand what is the goal of this PR. I can understand getting rid of obsolete macros in `configure.in` and updating the bundled `.m4` files, but clearly this does more than this. Why and what is the benefit?

File: .gitignore (hunk 6 above)
Comment (Apteryks) [reply .gitignore:25]: Okay. I'll drop this commit.

File: configure.in (hunk 7 above)
Comment (Apteryks) [reply configure.in:1349]: Will do!

**Comment by Apteryks** [comment]
Sorry for the long delay!

> 1. Why does `aclocal.m4` need to be removed? As long as we store all the generated files in the VCS (and, for better or worse, we do it), it should stay IMO.

Mostly out of my personal opinion that checking in things in VCS for no clear purpose is a bad practice.  If there's a purpose and it is explained, I'm open to revisit, of course.  Also the various M4 files of Autoconf versioned in the repo were clashing with a newer Autoconf, causing errors.

>     2. If we remove the "autohacks" stuff, wouldn't it result in configure generated under different systems be different again? Perhaps we don't care about it any longer because I personally just use `build/tools/autoconf/Dockerfile` for regenerating this anyhow, but this is not at all clear to me.

The autohacks prepending versioned M4 was causing issues with my recent Autoconf.  Perhaps it could have been updated, but it seems cleaner and simpler to just tell people to use Autoconf at version X instead of trying to check in Autoconf in the repo.

> 
> More generally, I'd really like to understand what is the goal of this PR.

Getting to gripe with this project's Autotools-based build system so that I could then extend it for the use cases I wanted it to support (using libraries from the system instead of bundling them), as I started to do with https://github.com/wxWidgets/wxWidgets/pull/23654.

I'll push a revised version fixing more whitespace problems shortly.

**Comment by Apteryks** [comment]
Hi, the https://github.com/wxWidgets/wxWidgets/actions/runs/5594401393/jobs/10229203621 error is likely a fluke, given all the other builds succeeded the test.  The whitespace check error is the upstream file ax_cxx_compile_stdcxx.m4; there is a tab mixed with spaces on a single line.  If we can't ignore it I can try to have upstream update it without the tab.

**Comment by vadz** [comment]
> > 1. Why does `aclocal.m4` need to be removed? As long as we store all the generated files in the VCS (and, for better or worse, we do it), it should stay IMO.
> 
> Mostly out of my personal opinion that checking in things in VCS for no clear purpose is a bad practice.

FWIW I used to agree with this but we've been doing this for ages in order to make things simpler for our users and I have to admit that in practice it is quite convenient because you can just tell people to run `./configure` instead of telling them to install autoconf and run it first, so I'm not so sure about this any more.

But, of course, the real argument against removing `aclocal.m4` is that we still keep tons of other generated files in the VCS -- so why remove just this one? This doesn't make any sense to me. There are 2 sensible choices: not store any generated files or store all of them. As I don't think there is any consensus about doing the former, could we please stick with the latter and keep `aclocal.m4` too?

> > 2. If we remove the "autohacks" stuff, wouldn't it result in configure generated under different systems be different again? Perhaps we don't care about it any longer because I personally just use `build/tools/autoconf/Dockerfile` for regenerating this anyhow, but this is not at all clear to me.
> 
> The autohacks prepending versioned M4 was causing issues with my recent Autoconf. Perhaps it could have been updated, but it seems cleaner and simpler to just tell people to use Autoconf at version X instead of trying to check in Autoconf in the repo.

OK, let's remove them, I've never really understood how this worked anyhow. But, as already mentioned previously, I still think using Docker container would be best as it's the only way to avoid gratuitous differences in `configure` which works in practice.

> > More generally, I'd really like to understand what is the goal of this PR.
> 
> Getting to gripe with this project's Autotools-based build system and modernizing/streamlining it a bit in the process so that I could then extend it for the use cases I wanted it to support (using libraries from the system instead of bundling them), as I started to do with #23654.

We already do the latter for many libraries (although not for CATCH, as we started using it back when it wasn't available as a system package anyhow), so it doesn't seem necessary to change much/anything in order to achieve this goal.

> I'll push a revised version fixing more whitespace problems shortly.

This is a minor problem, but if you look at the latest diff you can still easily see a few whitespace-only changes and the brackets placement is rather inconsistent even among the new lines added by this PR. It would be nice to standardize on some kind of consistent style.

To summarize:

1. I'd like to apply the changes to `configure.in` replacing deprecated macros with newer ones. If these changes could use consistent style, it would be ideal, but if not, we can live without it.
1. I am neutral about autohacks thing, i.e. I'm not especially eager to remove it, but we can do it if you'd like to.
1. I'd like to keep `aclocal.m4`.

Could it be possible to update the PR to do things like this? TIA!

**Comment by Apteryks** [comment]
Hello,

> * I'd like to apply the changes to `configure.in` replacing deprecated macros with newer ones. If these changes could use consistent style, it would be ideal, but if not, we can live without it.

I've already done a good effort on sanitizing the form, I don't see how much I can improve on it.  The new constructs involve nested macros which does change the aspect of the original, and it seems a common practice to leave the [[ ]] on their own line when double-quoting `#include <...>` to avoid confusing Emacs' autoconf-mode, possibly other editors. 

> * I am neutral about autohacks thing, i.e. I'm not especially eager to remove it, but we can do it if you'd like to.

Since it prevented me to run `make -f build/autogen.mk` using Autoconf 2.72 or 2.71, I think it's best to remove it.

> * I'd like to keep `aclocal.m4`.

You mentioned the discrepancy between keeping configure and other generated files vs aclocal.m4.  First, `configure` doesn't need `aclocal.m4`.  The configure script is wholly standalone.  That's the point of Autoconf.  Since Bakefile is used, to have a completely working build system checked in Git, its generated Makefiles must also be versioned.

The rest such as aclocal.m4, which is only needed by Autoconf at the time of *generating* `configure`, can be left to generate at that time, especially since it's automated in the documented process: `make -f build/autogen.mk`.  That was my rationale for getting rid of it.

**Comment by vadz** [comment]
Sorry, after finally spending literally 2 hours on this, I've given up and just redid some of what you wanted to do myself. Reasons:

1. The diff to `configure.in` is much more readable (i.e. can be actually reviewed) when doing the changes manually instead of letting `autoupdate` make them.
2. Updates to third party m4 files still resulted in warnings for `AC_TRY_XXX` used in them, which seems counterproductive, so I've just tweaked them myself to avoid the warnings too.
3. I kept `aclocal.m4` because I see no reason to remove it, as discussed.
4. I did remove autohacks even though I didn't have any problems with using them with 2.71, I suspect you might have not cleaned up the `.m4f` file and/or `autom4te.cache` when switching autoconf versions or something like this.
5. I also did something that I think is actually useful for the library users and got rid of `AM_PATH_GTK` stuff in favour of pkg-config.

I've taken the one commit from here that I could take easily, but I don't plan to apply any other ones, sorry. Of course, any testing of #23730 and other changes on top of it are still welcome, but it would be preferable to discuss them first and please avoid submitting changes done by automated tools. Thanks in advance!

File: autogen.sh
Comment (Apteryks) [inline autogen.sh:1]: @vadz I think this change to run`make -f build/autogen.mk` is better than just `aclocal`, as it builds the Bakefile-generated makefiles as well.

Comment (vadz) [reply autogen.sh:1]: We're going to stop using bakefile in hopefully not too distant future, so this doesn't matter much and in the meanwhile it seems better not to require that contributors have bakefile installed.

Comment (Apteryks) [reply autogen.sh:1]: OK! What's the long term plan?  Move to CMake altogether?  Or keep Autoconf as well and replace Bakefile with something else?

Comment (vadz) [reply autogen.sh:1]: I don't know :-( Personally I'd like to have something other than CMake but there is nothing that would work out of the box and I don't know if there is anything that would work at all for us. Please search for "premake" in wx-dev archives for more.
